minecraft/protocol/packet/compression.go

```diff
@@ -106,20 +108,43 @@ func (flateCompression) Compress(decompressed []byte) ([]byte, error) {
 
 // Decompress ...
 func (flateCompression) Decompress(compressed []byte, limit int) ([]byte, error) {
-	buf := bytes.NewReader(compressed)
-	c := flateDecompressPool.Get().(io.ReadCloser)
-	defer flateDecompressPool.Put(c)
+	r := flateDecompressPool.Get().(io.ReadCloser)
+	defer func() {
+		_ = r.Close()
+		flateDecompressPool.Put(r)
+	}()
 
-	if err := c.(flate.Resetter).Reset(buf, nil); err != nil {
+	if err := r.(flate.Resetter).Reset(bytes.NewReader(compressed), nil); err != nil {
 		return nil, fmt.Errorf("reset flate: %w", err)
 	}
-	_ = c.Close()
 
-	// Guess an uncompressed size of 2*len(compressed).
-	decompressed := bytes.NewBuffer(make([]byte, 0, len(compressed)*2))
-	if _, err := io.Copy(decompressed, io.LimitReader(c, int64(limit))); err != nil {
+	var decompressed bytes.Buffer
+	// If the compressed data is less than half the limit, we can safely assume l*2, otherwise cap at limit.
+	l := len(compressed)
+	capHint := limit
+	if l <= limit/2 {
+		capHint = l * 2
+	}
+	decompressed.Grow(capHint)
+
+	// Handle no limit
+	if limit == math.MaxInt {
+		if _, err := io.Copy(&decompressed, r); err != nil {
+			return nil, fmt.Errorf("decompress flate: %w", err)
+		}
+		return decompressed.Bytes(), nil
+	}
```

Comment:
Is there ever a sensible case to have no limit?

Reply:
limit + 1 would cause an overflow so we cant handle this without checking limit == math.MaxInt, which is what the listener sets if the user configures no limit
https://github.com/Sandertv/gophertunnel/blob/8096f00d11d6d7ee3472b59fc443dfdf6911e29e/minecraft/listener.go#L133